File: apps/web/src/lib/api/invite.ts
/**
 * Invite API
 * ----------
 * Typed wrappers for all /invites endpoints.
 * Mirrors the Pydantic schemas from apps/backend/app/api/schemas/invite.py.
 *
 * Phone numbers are always masked in responses (+91****5592).
 * The backend handles encryption/decryption — the frontend never sees raw phones.
 */

import { apiClient } from "./client";
import type { OffsetPageMeta } from "./community";
import { toQueryString } from "./utils";

/* ══════════════════════════════════════════════════════════════════════
   ENUMS
   ══════════════════════════════════════════════════════════════════════ */

export type InviteStatus = "pending" | "accepted" | "rejected" | "expired" | "revoked";

export type InviteRelation =
  | "Family Member"
  | "Doctor"
  | "Caregiver"
  | "Lab / Diagnostic"
  | "Pharmacy"
  | "Other";

export type InviteAccessScope = "App Access" | "Group Access";

export type InviteDirection = "sent" | "received";

/* ══════════════════════════════════════════════════════════════════════
   REQUEST TYPES
   ══════════════════════════════════════════════════════════════════════ */

export interface SendInviteData {
  phone: string;
  invitee_name?: string;
  relation: InviteRelation;
  access_scope?: InviteAccessScope;
  message?: string;
}

export interface RejectInviteData {
  reason?: string;
}

export interface InviteListParams {
  direction?: InviteDirection;
  status?: InviteStatus;
  page?: number;
  page_size?: number;
}

/* ══════════════════════════════════════════════════════════════════════
   RESPONSE TYPES
   ══════════════════════════════════════════════════════════════════════ */

export interface InviterOut {
  id: string;
  name: string;
  initials: string;
  phone_masked: string;
}

export interface InviteeOut {
  name: string | null;
  phone_masked: string;
}

export interface InviteOut {
  id: string;
  inviter: InviterOut;
  invitee: InviteeOut;
  relation: InviteRelation;
  access_scope: InviteAccessScope;
  status: InviteStatus;
  direction: InviteDirection;
  message: string | null;
  created_at: string;
  updated_at: string | null;
  expires_at: string | null;
  accepted_at: string | null;
  rejected_at: string | null;
  group_id: string | null;
}

export interface InviteDetailOut extends InviteOut {
  group_name: string | null;
  group_slug: string | null;
}

export interface SendInviteResponse {
  success: boolean;
  invite: InviteOut;
  message: string;
}

export interface InviteActionResponse {
  success: boolean;
  invite: InviteOut;
  message: string;
}

export interface InviteListResponse {
  items: InviteOut[];
  meta: OffsetPageMeta;
}

export interface InviteCountsOut {
  sent_pending: number;
  sent_total: number;
  received_pending: number;
  received_total: number;
}

/* ── Lookup phone (invite modal — verify against store) ── */
export interface LookupPhoneData {
  phone: string;
}

export interface LookupPhoneResponse {
  registered: boolean;
  phone_masked: string;
  name: string | null;
}

/* ── Register new user via invite OTP flow ── */
export interface RegisterInviteeData {
  phone: string;
  code: string;
  name?: string;
}

export interface RegisterInviteeResponse {
  success: boolean;
  user_id: string;
  phone_masked: string;
  name: string;
  message: string;
}

/* ══════════════════════════════════════════════════════════════════════
   API CLIENT
   ══════════════════════════════════════════════════════════════════════ */

export const inviteApi = {
  /* ── Send invite ── */
  sendInvite: (data: SendInviteData) =>
    apiClient<SendInviteResponse>("/invites", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  /* ── List invites ── */
  listInvites: (params: InviteListParams = {}) =>
    apiClient<InviteListResponse>(`/invites${toQueryString(params)}`),

  /* ── Invite counts ── */
  getCounts: () =>
    apiClient<InviteCountsOut>("/invites/counts"),

  /* ── Get invite detail ── */
  getInvite: (inviteId: string) =>
    apiClient<InviteDetailOut>(`/invites/${inviteId}`),

  /* ── Accept invite ── */
  acceptInvite: (inviteId: string) =>
    apiClient<InviteActionResponse>(`/invites/${inviteId}/accept`, {
      method: "POST",
      body: JSON.stringify({}),
    }),

  /* ── Reject invite ── */
  rejectInvite: (inviteId: string, data: RejectInviteData = {}) =>
    apiClient<InviteActionResponse>(`/invites/${inviteId}/reject`, {
      method: "POST",
      body: JSON.stringify(data),
    }),

  /* ── Resend invite ── */
  resendInvite: (inviteId: string) =>
    apiClient<InviteActionResponse>(`/invites/${inviteId}/resend`, {
      method: "POST",
    }),

  /* ── Revoke invite ── */
  revokeInvite: (inviteId: string) =>
    apiClient<InviteActionResponse>(`/invites/${inviteId}`, {
      method: "DELETE",
    }),

  /* ── Lookup phone — does this number already exist on ArogyaVault? ── */
  lookupPhone: (data: LookupPhoneData) =>
    apiClient<LookupPhoneResponse>("/invites/lookup", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  /* ── Register a brand-new user (dev OTP = 123456) ── */
  registerInvitee: (data: RegisterInviteeData) =>
    apiClient<RegisterInviteeResponse>("/invites/register-user", {
      method: "POST",
      body: JSON.stringify(data),
    }),
};

File: apps/backend/app/api/schemas/invite.py
"""
Pydantic schemas for invite flow endpoints.

Industry-standard patterns:
  - Phone numbers stored encrypted (AES-256-GCM) — never in plaintext
  - HMAC-SHA256 hash for dedup lookups without decryption
  - Masked phone for UI display (+91****5592)
  - Status lifecycle: pending → accepted/rejected/expired/revoked
  - Offset-paginated invite lists with filters
"""

from __future__ import annotations

from enum import Enum
from pydantic import BaseModel, Field

from app.api.schemas.community import OffsetPageMeta


# ══════════════════════════════════════════════════════════════════════════════
#  ENUMS
# ══════════════════════════════════════════════════════════════════════════════


class InviteStatus(str, Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    REJECTED = "rejected"
    EXPIRED = "expired"
    REVOKED = "revoked"


class InviteRelation(str, Enum):
    FAMILY = "Family Member"
    DOCTOR = "Doctor"
    CAREGIVER = "Caregiver"
    LAB = "Lab / Diagnostic"
    PHARMACY = "Pharmacy"
    OTHER = "Other"


class InviteAccessScope(str, Enum):
    APP_ACCESS = "App Access"         # Full app-level sharing
    GROUP_ACCESS = "Group Access"     # Limited to a specific group


class InviteDirection(str, Enum):
    SENT = "sent"       # Current user sent the invite
    RECEIVED = "received"  # Current user received the invite


# ══════════════════════════════════════════════════════════════════════════════
#  REQUEST SCHEMAS
# ══════════════════════════════════════════════════════════════════════════════


class SendInviteRequest(BaseModel):
    """Send a new invite — phone is encrypted before storage."""
    phone: str = Field(
        ...,
        examples=["+919876543210"],
        description="Invitee phone in E.164 format",
    )
    invitee_name: str | None = Field(
        None,
        max_length=100,
        description="Optional display name for the invitee",
    )
    relation: InviteRelation = Field(
        ...,
        description="Relationship type with the invitee",
    )
    access_scope: InviteAccessScope = Field(
        default=InviteAccessScope.GROUP_ACCESS,
        description="Level of access to grant",
    )
    message: str | None = Field(
        None,
        max_length=500,
        description="Optional personal message with the invite",
    )


class ResendInviteRequest(BaseModel):
    """Resend an existing pending invite (resets expiry)."""
    pass  # invite_id comes from path parameter


class AcceptInviteRequest(BaseModel):
    """Accept a received invite."""
    pass  # invite_id comes from path parameter


class RejectInviteRequest(BaseModel):
    """Reject a received invite."""
    reason: str | None = Field(
        None,
        max_length=200,
        description="Optional reason for declining",
    )


# ══════════════════════════════════════════════════════════════════════════════
#  RESPONSE SCHEMAS
# ══════════════════════════════════════════════════════════════════════════════


class InviterOut(BaseModel):
    """Who sent the invite."""
    id: str
    name: str
    initials: str
    phone_masked: str  # +91****5592


class InviteeOut(BaseModel):
    """Who received the invite (phone is always masked)."""
    name: str | None = None
    phone_masked: str  # +91****5592


class InviteOut(BaseModel):
    """Single invite — returned in lists and detail views."""
    id: str
    inviter: InviterOut
    invitee: InviteeOut
    relation: InviteRelation
    access_scope: InviteAccessScope
    status: InviteStatus
    direction: InviteDirection  # relative to the requesting user
    message: str | None = None
    created_at: str
    updated_at: str | None = None
    expires_at: str | None = None
    accepted_at: str | None = None
    rejected_at: str | None = None
    group_id: str | None = None  # set after acceptance (linked group created)


class InviteDetailOut(InviteOut):
    """Extended detail — includes the linked group info if accepted."""
    group_name: str | None = None
    group_slug: str | None = None


class SendInviteResponse(BaseModel):
    """Response after sending an invite."""
    success: bool = True
    invite: InviteOut
    message: str = "Invite sent successfully"


class InviteActionResponse(BaseModel):
    """Response for accept/reject/revoke/resend actions."""
    success: bool = True
    invite: InviteOut
    message: str


class InviteListResponse(BaseModel):
    """Paginated list of invites."""
    items: list[InviteOut]
    meta: OffsetPageMeta


class InviteCountsOut(BaseModel):
    """Quick counts for the invite dashboard."""
    sent_pending: int = 0
    sent_total: int = 0
    received_pending: int = 0
    received_total: int = 0


# ══════════════════════════════════════════════════════════════════════════════
#  LOOKUP + REGISTER (invite modal flow)
# ══════════════════════════════════════════════════════════════════════════════


class LookupPhoneRequest(BaseModel):
    """Check whether a phone is already registered on ArogyaVault."""
    phone: str = Field(..., description="Phone number in E.164 or loose format")
    group_id: str | None = Field(
        default=None,
        description=(
            "Optional group id the invite modal is currently scoped to. "
            "When supplied, the response will set already_member=True if the "
            "looked-up phone already belongs to this group so the modal can "
            "short-circuit into a cancel-only state."
        ),
    )


class LookupPhoneResponse(BaseModel):
    """Lookup result — safe for the invite modal."""
    registered: bool
    phone_masked: str
    name: str | None = None
    already_member: bool = Field(
        default=False,
        description=(
            "True when the looked-up phone is already a member of the "
            "group_id passed in the request. The invite modal uses this "
            "to show an 'already a member' state with only a Cancel button."
        ),
    )


class RegisterInviteeRequest(BaseModel):
    """
    Register a brand-new user via the invite OTP flow (dev OTP = 123456).

    `invite_level` drives the post-registration behaviour:
      - "app"         → register only, no group created
      - "group"       → register + create a fresh linked group between inviter+invitee
      - "<group_id>"  → register + add invitee to an existing group owned by the inviter
    """
    phone: str = Field(..., description="Invitee phone in E.164 format")
    code: str = Field(..., description="OTP verification code")
    name: str | None = Field(None, max_length=100)
    invite_level: str = Field(
        default="group",
        description="'app', 'group', or a group id the inviter wants to reuse",
    )
    relation: InviteRelation | None = Field(
        default=InviteRelation.FAMILY,
        description="Relationship to record on the newly-created linked group",
    )
    access_scope: InviteAccessScope | None = Field(
        default=InviteAccessScope.APP_ACCESS,
        description="Access scope for the new linked group",
    )


class RegisterInviteeResponse(BaseModel):
    success: bool = True
    user_id: str
    phone_masked: str
    name: str
    group_id: str | None = None
    message: str = "Invitee registered successfully"
